autopilot/contractor/contractor.go

```diff
@@ -745,212 +625,178 @@ func renterFundsToExpectedStorage(renterFunds types.Currency, duration uint64, p
 			} else {
 				logger.Debug("successfully archived contract")
+				usability.Infof("updated contract usability from %s -> bad, reason: %v", c.Usability, reason)
 			}
-			churnReasons[c.ID] = reason.Error()
 			continue
 		}
 
 		// fetch host
 		host, err := bus.Host(ctx, c.HostKey)
 		if err != nil {
 			logger.With(zap.Error(err)).Warn("missing host")
-			churnReasons[c.ID] = api.ErrUsabilityHostNotFound.Error()
+			updateUsability(ctx, cm, api.ContractUsabilityBad, api.ErrUsabilityHostNotFound.Error(), usability)
 			continue
 		}
 
 		// extend logger
-		logger = logger.With("addresses", host.ResolvedAddresses).
+		logger = logger.
+			With("addresses", host.ResolvedAddresses).
 			With("blocked", host.Blocked)
 
 		// check if host is blocked
 		if host.Blocked {
 			logger.Info("host is blocked")
-			churnReasons[c.ID] = api.ErrUsabilityHostBlocked.Error()
+			updateUsability(ctx, cm, api.ContractUsabilityBad, api.ErrUsabilityHostBlocked.Error(), usability)
 			continue
 		}
 
 		// check if host has a redundant ip
 		if ctx.ShouldFilterRedundantIPs() && ipFilter.HasRedundantIP(host) {
 			logger.Info("host has redundant IP")
-			churnReasons[c.ID] = api.ErrUsabilityHostRedundantIP.Error()
+			updateUsability(ctx, cm, api.ContractUsabilityBad, api.ErrUsabilityHostRedundantIP.Error(), usability)
 			continue
 		}
 
 		// get check
 		if host.Checks == (api.HostChecks{}) {
 			logger.Warn("missing host check")
-			churnReasons[c.ID] = api.ErrUsabilityHostNotFound.Error()
+			updateUsability(ctx, cm, api.ContractUsabilityBad, api.ErrUsabilityHostCheckNotFound.Error(), usability)
 			continue
 		}
 
-		// NOTE: if we have a contract with a host that is not scanned, we either
-		// added the host and contract manually or reset the host scans. In that case,
-		// we ignore the fact that the host is not scanned for now to avoid churn.
-		if inSet && host.Checks.UsabilityBreakdown.NotCompletingScan {
-			keepContract(c.ContractMetadata, host)
+		// NOTE: if we have a contract with a host that is not scanned, we
+		// either added the host and contract manually or reset the host scans.
+		// In that case, we ignore the fact that the host is not scanned for now
+		// to avoid churn.
+		if c.IsGood() && host.Checks.UsabilityBreakdown.NotCompletingScan {
+			ipFilter.Add(host)
 			logger.Info("ignoring contract with unscanned host")
+			updateUsability(ctx, cm, api.ContractUsabilityGood, "not completing scan", usability)
 			continue // no more checks until host is scanned
 		}
 
 		// check usability
 		if !host.Checks.UsabilityBreakdown.IsUsable() {
 			reasons := strings.Join(host.Checks.UsabilityBreakdown.UnusableReasons(), ",")
 			logger.With("reasons", reasons).Info("unusable host")
-			churnReasons[c.ID] = reasons
+			updateUsability(ctx, cm, api.ContractUsabilityBad, reasons, usability)
 			continue
 		}
 
 		// check if revision is available
 		if c.Revision == nil {
-			if inSet && remainingLeeway > 0 {
```

Comment:
Why get rid of the leeway?

Reply:
With contract sets we had a desired number of contracts and would exclude contracts from that set if we already had plenty of other (better) ones. Now that we removed contract sets I feel things are very different now because we don't mark contracts as bad at a certain point. There were a couple of smaller things that triggered me too, it's a `leeway`, the docstring mentions `maxKeepLeeway` which we no longer have and it felt like complexity we don't need (anymore).

Reply:
> because we don't mark contracts as bad at a certain point

Can you expand on this? In theory we want the exact same behavior now that we would have with a single contract set. A contract that is "bad" was essentially just moved out of the contract set and will lead to migrations to avoid losing data when it doesn't get renewed and expires.
To avoid that on temporary network errors we added a leeway. So this shouldn't be obsolete now.

Reply:
I don't mind re-adding the leeway at all. I removed it because it was yet another leeway, but mostly because I feel we added it back when we had `siad` hosts that would time out on occasion when fetching the revision. I say that because we had a host revision timeout of 4 minutes (I believe) because we had EA withrawal timeouts at one point. I kind of want to remove it, reassess the churn and go from there. I just want to question these fields going into the refactor and RHP4.